File: src/python/xerxes/training/rl/envs.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""RL environment registry plus a few built-in environment definitions.

Each env carries metadata and an optional ``reward_fn`` callable
that maps a trajectory dict to a scalar reward. Real RL workloads
(SWE-bench, terminal-test, agentic-OPD) live in ``/environments/``
outside the package; this module is the discovery surface Xerxes
exposes to RL drivers.
"""

from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

RewardFn = Callable[[dict[str, Any]], float]


@dataclass
class RLEnvironment:
    """Metadata describing one RL training environment.

    Attributes:
        name: unique slug used for registration and CLI lookup.
        description: short human-readable summary.
        config_path: optional path to an env-specific YAML/JSON.
        backend: ``"local"`` / ``"modal"`` / ``"docker"`` / etc.
        tags: free-form labels for filtering.
        reward_fn: trajectory -> reward callable; absent for envs
            scored entirely by the backend.
    """

    name: str
    description: str
    config_path: str = ""
    backend: str = "local"
    tags: list[str] = field(default_factory=list)
    reward_fn: RewardFn | None = None


class RLEnvironmentRegistry:
    """In-process registry mapping env name to :class:`RLEnvironment`."""

    def __init__(self) -> None:
        """Start with an empty registry."""
        self._envs: dict[str, RLEnvironment] = {}

    def register(self, env: RLEnvironment) -> None:
        """Register ``env``; raises ``ValueError`` if its name is empty."""
        if not env.name:
            raise ValueError("env name required")
        self._envs[env.name] = env

    def list_envs(self) -> list[RLEnvironment]:
        """Return all registered environments sorted by name."""
        return sorted(self._envs.values(), key=lambda e: e.name)

    def get(self, name: str) -> RLEnvironment | None:
        """Look up an environment by name, or ``None`` if absent."""
        return self._envs.get(name)


def builtin_envs() -> RLEnvironmentRegistry:
    """Return a registry pre-populated with the bundled environments."""
    reg = RLEnvironmentRegistry()
    reg.register(
        RLEnvironment(
            name="xerxes-terminal-test",
            description="Synthetic 4-task curriculum (greeting.txt, count.txt, answer.txt, eval-arithmetic)",
            backend="local",
            tags=["coding", "tooling"],
            reward_fn=lambda traj: 1.0 if traj.get("passed_all_tasks") else 0.0,
        )
    )
    reg.register(
        RLEnvironment(
            name="xerxes-swe-bench",
            description="SWE-bench style env: write code → tests verify reward.",
            backend="modal",
            tags=["coding", "swe"],
            reward_fn=lambda traj: float(traj.get("tests_passed", 0)) / max(1.0, float(traj.get("tests_total", 1))),
        )
    )
    reg.register(
        RLEnvironment(
            name="agentic-opd",
            description="Objective Policy Distribution: open-ended objective scoring.",
            backend="modal",
            tags=["research"],
        )
    )
    return reg


__all__ = ["RLEnvironment", "RLEnvironmentRegistry", "RewardFn", "builtin_envs"]

File: src/typescript/src/training/rl/envs.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

/** One completed rollout or evaluation passed to an environment reward function. */
export type RewardTrajectory = Readonly<Record<string, unknown>>

/** Maps a trajectory into a scalar reward supplied by the environment author. */
export type RewardFn = (trajectory: RewardTrajectory) => number

export interface RLEnvironmentOptions {
  /** Backend-specific configuration artifact, when the environment has one. */
  readonly configPath?: string
  readonly backend?: string
  readonly description: string
  readonly name: string
  readonly rewardFn?: RewardFn
  readonly tags?: readonly string[]
}

/** Immutable metadata and reward boundary for one reinforcement-learning environment. */
export class RLEnvironment {
  readonly backend: string
  readonly configPath: string
  readonly description: string
  readonly name: string
  readonly rewardFn: RewardFn | undefined
  readonly tags: readonly string[]

  constructor(options: RLEnvironmentOptions) {
    this.name = requiredText(options.name, 'name')
    this.description = text(options.description, 'description')
    this.configPath = text(options.configPath ?? '', 'configPath')
    this.backend = text(options.backend ?? 'local', 'backend')
    this.tags = Object.freeze((options.tags ?? []).map((tag, index) => requiredText(tag, `tags[${index}]`)))
    this.rewardFn = options.rewardFn
    Object.freeze(this)
  }

  /** Evaluate a trajectory when this environment owns scoring. */
  reward(trajectory: RewardTrajectory): number | undefined {
    const rewardFn = this.rewardFn
    if (rewardFn === undefined) return undefined
    const value = rewardFn({ ...trajectory })
    if (!Number.isFinite(value)) throw new TypeError(`environment ${this.name} returned a non-finite reward`)
    return value
  }
}

/** In-process registry for environment metadata supplied by the embedding host. */
export class RLEnvironmentRegistry {
  private readonly environments = new Map<string, RLEnvironment>()

  /** Register or replace an environment by its stable name. */
  register(environment: RLEnvironment): void {
    if (!(environment instanceof RLEnvironment)) throw new TypeError('environment must be an RLEnvironment')
    this.environments.set(environment.name, environment)
  }

  /** Return registered environments in deterministic name order. */
  listEnvironments(): readonly RLEnvironment[] {
    return [...this.environments.values()].sort((left, right) => left.name.localeCompare(right.name))
  }

  /** Look up one environment without manufacturing an external backend result. */
  get(name: string): RLEnvironment | undefined {
    return this.environments.get(name)
  }
}

/** Construct the small built-in environment catalogue available without a provider SDK. */
export function builtinEnvironments(): RLEnvironmentRegistry {
  const registry = new RLEnvironmentRegistry()
  registry.register(new RLEnvironment({
    name: 'xerxes-terminal-test',
    description: 'Synthetic 4-task curriculum (greeting.txt, count.txt, answer.txt, eval-arithmetic)',
    backend: 'local',
    tags: ['coding', 'tooling'],
    rewardFn: trajectory => trajectory.passed_all_tasks === true ? 1 : 0,
  }))
  registry.register(new RLEnvironment({
    name: 'xerxes-swe-bench',
    description: 'SWE-bench style env: write code → tests verify reward.',
    backend: 'modal',
    tags: ['coding', 'swe'],
    rewardFn: trajectory => {
      const passed = finiteNumber(trajectory.tests_passed) ?? 0
      const total = finiteNumber(trajectory.tests_total) ?? 1
      return passed / Math.max(1, total)
    },
  }))
  registry.register(new RLEnvironment({
    name: 'agentic-opd',
    description: 'Objective Policy Distribution: open-ended objective scoring.',
    backend: 'modal',
    tags: ['research'],
  }))
  return registry
}

function requiredText(value: string, name: string): string {
  const normalized = text(value, name).trim()
  if (!normalized) throw new TypeError(`${name} must be non-empty`)
  return normalized
}

function text(value: string, name: string): string {
  if (typeof value !== 'string') throw new TypeError(`${name} must be a string`)
  return value
}

function finiteNumber(value: unknown): number | undefined {
  return typeof value === 'number' && Number.isFinite(value) ? value : undefined
}
